### PR TITLE
add example of qml, add qml class description

### PR DESCRIPTION
**Pull request by dorotaphanSiili**
-method name start with lower first
-signal name starts with lower letter
-api factor for interface returns the shared_ptr<> instead of raw ptr
-change olink factory to get Client in constructor, and use it to link created interfaceClient on return
-added WRITE to QPROPERY in qml wrapper
-fixed a bug, where olink server added ")" to property and signal name

### CODE REVIEW

File: goldenmaster/tb_enum/olink/olinkfactory.h [hunk 1]
```diff
@@ -13,14 +14,23 @@ namespace tb_enum {
 * You can check the usage of ApiFactoryInterface in Qml versions of interface implementation.
 * By default this factory is not used as ApiFactoryInterface.
 * See also the ApiFactory, where you can set this factory as an ApiFactoryInterface implementation.
+*
+* WARNING when creating the interface, make sure that server side already has the source (the interface service).
+*    OR use unconnected client when creating the interface and connect it the source object is already there.
+*
+* WARNING Because the registry held by client cannot have multiple objects with same objectId (which is the combination of module and interface)
+*         you will be able to have only one instance of each of your Interfaces connected to client at one time.
+*         other interafce instances wont be linked
```
Comment (w4bremer) [inline goldenmaster/tb_enum/olink/olinkfactory.h:23]: type interface

File: goldenmaster/examples/qml/main.qml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    Tb_same2SameStruct1Interface { id: qmlTb_same2SameStruct1Interface }
+    Tb_same2SameStruct2Interface { id: qmlTb_same2SameStruct2Interface }
+    Tb_same2SameEnum1Interface { id: qmlTb_same2SameEnum1Interface }
+    Tb_same2SameEnum2Interface { id: qmlTb_same2SameEnum2Interface }
+    Tb_simpleSimpleInterface { id: qmlTb_simpleSimpleInterface }
+    Tb_simpleSimpleArrayInterface { id: qmlTb_simpleSimpleArrayInterface }
+    Testbed1StructInterface { id: qmlTestbed1StructInterface }
+    Testbed1StructArrayInterface { id: qmlTestbed1StructArrayInterface }
+
+        Button {
```
Comment (w4bremer) [inline goldenmaster/examples/qml/main.qml:38]: indentation seems to be off by one level

**Review by w4bremer — APPROVED** [review]
Minor issues, but looks good

File: goldenmaster/examples/qml/main.qml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ApplicationWindow {
+    id: appWindow
+    visible: true
+    width: 300
+    height: 300
+    ColumnLayout {
+        spacing: 10
+        id: mainLayout
+        anchors.fill: parent
+    Testbed2ManyParamInterface { id: qmlTestbed2ManyParamInterface }
```
Comment (jryannel) [inline goldenmaster/examples/qml/main.qml:20]: the interfaces should be on the root level at the end. So the visual aspects are shown first.
```qml
ApplicationWindow {
  ColumnLayout { ... }
  // interfaces ...
}
```

File: goldenmaster/tb_enum/plugin/qmldir [hunk 4]
```diff
@@ -1,2 +1,2 @@
-module tb.enum
+module tb_enum
```
Comment (jryannel) [inline goldenmaster/tb_enum/plugin/qmldir:1]: why using `_` instead of dot? using a dot is more qml like.

File: templates/api/api.h.tpl [hunk 5]
```diff
@@ -170,11 +172,12 @@ public:
 {{- end }}   {{- /* end range operation param*/}}  
     */
 {{- end }}   {{- /* end if operations description */}}
-    virtual {{qtReturn "" .Return }} {{.Name}}({{qtParams "" .Params}}) = 0;
+    virtual {{qtReturn "" .Return }} {{lower1 .Name}}({{qtParams "" .Params}}) = 0;
```
Comment (jryannel) [inline templates/api/api.h.tpl:175]: you could use `{{ camel .Name }} which makes camel case with lower 1st. or `{{ qtVar .Name }}`

File: templates/api/apifactory.cpp.tpl [hunk 6]
```diff
@@ -30,9 +30,9 @@ ApiFactoryInterface * ApiFactory::get()
 
 {{- range .Module.Interfaces }}
 
-Abstract{{.Name}}* ApiFactory::create{{Camel .Name}}(QObject *parent) 
+std::shared_ptr<Abstract{{Camel .Name}}> ApiFactory::create{{Camel .Name}}(QObject *parent) 
 {
-    return new Simulation{{.Name}}(parent);
+    return std::make_shared<Simulation{{.Name}}>(parent);
```
Comment (jryannel) [inline templates/api/apifactory.cpp.tpl:35]: Maybe use {{ Camel .Name }}

File: templates/api/simu.cpp.tpl [hunk 7]
```diff
@@ -76,7 +76,7 @@ void {{$class}}::set{{Camel .Name}}({{qtParam "" .}})
 {{- end }}
 {{- range .Operations }}
 
-{{qtReturn "" .Return}} {{$class}}::{{.Name}}({{qtParams "" .Params}})
+{{qtReturn "" .Return}} {{$class}}::{{lower1 .Name}}({{qtParams "" .Params}})
```
Comment (jryannel) [inline templates/api/simu.cpp.tpl:79]: again using `{{ camel .Name }}`

File: templates/api/simu.h.tpl [hunk 8]
```diff
@@ -26,7 +26,7 @@ public:
 {{- end }}
 
 {{- range .Operations }}
-    {{qtReturn "" .Return}} {{.Name}}({{qtParams "" .Params}}) override;
+    {{qtReturn "" .Return}} {{lower1 .Name}}({{qtParams "" .Params}}) override;
```
Comment (jryannel) [inline templates/api/simu.h.tpl:29]: `{{ camel .Name}}`

File: templates/examples/qml/main.qml.tpl [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    id: appWindow
+    visible: true
+    width: 300
+    height: 300
+    ColumnLayout {
+        spacing: 10
+        id: mainLayout
+        anchors.fill: parent
+
+{{- range.System.Modules }}
```
Comment (jryannel) [inline templates/examples/qml/main.qml.tpl:20]: move this to the end of the application window. This leaves the visual elements together.

**Review by jryannel — COMMENTED** [review]
often you can replace lower1 with camel. Typical qml modules use the dot separator. The main.qml should be structured that the visual parts come first and the interfaces are listed at the end.

File: goldenmaster/tb_enum/plugin/qmldir (hunk 4 above)
Comment (dorotaphanSiili) [reply goldenmaster/tb_enum/plugin/qmldir:1]: as I recall the qt didn't allowed me to import in qml with the dot

File: templates/api/api.h.tpl (hunk 5 above)
Comment (dorotaphanSiili) [reply templates/api/api.h.tpl:175]: I can change if that is what we want, but camel seems to do more, this one emphasizes that we change only one letter  🤔 
the qtVar is less readable so I would stick with lower1 or camel (if we want say we change more)

**Pull request merged**